File: examples/guests/python/transformer/transformer.py
from generated.guest.contracts import (
    TRANSFORMERDataTransformerPlugin,
    set_transformer_factory,
    polyplug_init,
)


class TransformerImpl(TRANSFORMERDataTransformerPlugin):
    """The factory receives the HostApi pointer at polyplug_init time."""

    def __init__(self, host_ptr: int) -> None:
        # Host handle for this runtime, captured at construction.
        self._host_ptr: int = host_ptr

    def transform(self, input: str) -> str:
        s = input.removeprefix("DECODED:")
        parts = s.split("|")
        if len(parts) >= 3:
            name = parts[0].upper()
            value = f"{parts[1]} (transformed)"
            count = int(parts[2]) + 1
            return f"TRANSFORMED:{name}|{value}|{count}"
        return "INVALID:format"


# Register the factory; the generated polyplug_init constructs the
# implementation with its owning runtime's host pointer.
set_transformer_factory(TransformerImpl)

File: examples/guests/csharp/transformer/Transformer.cs
using System.Runtime.CompilerServices;
using Polyplug.Guest;
using Polyplug.Abi;

namespace Transformer;

public sealed class TransformerPlugin : IDataTransformerGuestContract
{
    // Host handle for this runtime, captured at instance creation.
    private readonly IntPtr _host;

    public TransformerPlugin(IntPtr host)
    {
        _host = host;
    }

    public StringView Transform(StringView input)
    {
        string s = StringViewHelper.StripPrefix(input, "DECODED:");
        string[] parts = s.Split('|');
        if (parts.Length >= 3 && int.TryParse(parts[2], out int count))
        {
            string name = parts[0].ToUpper();
            string value = $"{parts[1]} (transformed)";
            return PolyplugHost.AllocString(_host, $"TRANSFORMED:{name}|{value}|{count + 1}");
        }
        return PolyplugHost.AllocString(_host, "INVALID:format");
    }
}

public static class Registration
{
    [ModuleInitializer]
    public static void Register()
    {
        // The factory receives the HostApi pointer per created instance.
        TransformerInterfaces.SetTransformerFactory(host => new TransformerPlugin(host));
    }
}
